### PR TITLE
ply: Initial support for riscv64

### LINKED ISSUE #81: Support for riscv64 architecture
It would be great to have riscv64 support in ply as this is a very lightweight BPF implementation. I'll try to get initial support working and open a PR.

cc @dlan17

### PR DESCRIPTION
**Pull request by xmzzz**
Add riscv64 support in ply, and makes BEGIN/END probes more reliable by implementing BPF_PROG_TYPE_RAW_TRACEPOINT probes and then calling BPF_PROG_TEST_RUN on them.

Tested on x86 and riscv64, and it works fine.

Resolves #81. Please take a look, thanks. @wkz @namhyung

Signed-off-by: Mingzheng Xing xingmingzheng@iscas.ac.cn

### CODE REVIEW

**Comment by wkz** [comment]
Great work!

There seems to be some backwards compatibility issues around the `s/read/read_kernel/` stuff. I'm leaning towards just requiring the existence of the `read_kernel` variants along with `bpf_prog_test_run` (from the looks of it that means at least Linux 5.5)

In that case, we can remove the old begin/end implementation, which would be nice.

I will work on a new test harness to that this can all be validated.

**Comment by xmzzz** [comment]
Great suggestion, I added two new commits, thanks for your review. @wkz

**Comment by xmzzz** [comment]
Hi, @wkz . I added new updates, including solving issue #83 .
Do you feel like there are too many commits here? Or I can split it into multiple PR if needed.

File: src/libply/built-in/memory.c [hunk 1]
```diff
@@ -141,11 +141,7 @@ static int str_ir_post(const struct func *func, struct node *n,
 	ir_emit_insn(ir, MOV_IMM((int32_t)type_sizeof(n->sym->type)), BPF_REG_2, 0);
 	ir_emit_sym_to_reg(ir, BPF_REG_3, ptr->sym);
 
-#if defined(__riscv) && __riscv_xlen == 64
         ir_emit_insn(ir, CALL(BPF_FUNC_probe_read_kernel_str), 0, 0); 
-#else
-        ir_emit_insn(ir, CALL(BPF_FUNC_probe_read_str), 0, 0); 
-#endif /* defined(__riscv) && __riscv_xlen == 64 */
```
Comment (namhyung) [inline src/libply/built-in/memory.c]: I think it's better to squash this commit to the origin change in the second commit.  And please mention it requires 5.5+ kernel.

**Comment by wkz** [comment]
I think we can keep it as one PR. But it would be great if you could squash some of the commits together to make review easier. Something like:
1. Use new `read_kernel` variants
2. Replace BEGIN/END implementation with `bpf_prog_test_run`
3. Add `riscv64` support

**Comment by xmzzz** [comment]
Hi, @wkz , @namhyung , thanks for your reply.
Accordingly, I reorganized my commit and it looks better now, please check it out.
Thanks.

**Comment by xmzzz** [comment]
This is my test result under three platforms of x86, riscv64, qemu riscv64, everything looks normal.

```
$ cat kretprobe.ply 
BEGIN {
  printf("BEGIN\n");
}

kretprobe:vfs_read {
  @["size"] = quantize(retval);
}

END {
  printf("END\n");
}
```

x86:

```
$ uname -a
Linux xmz-huawei 5.15.0-53-generic #59-Ubuntu SMP Mon Oct 17 18:53:30 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
$ sudo ply -c 'sleep 5' kretprobe.ply 
BEGIN

@:
{ size    }: 
	         < 0	     326 ┤██▌                             │
	[   0,    1]	     390 ┤███                             │
	[   2,    3]	      34 ┤▎                               │
	[   4,    7]	       5 ┤                                │
	[   8,   15]	    3087 ┤███████████████████████▍        │
	[  16,   31]	       4 ┤                                │
	[  32,   63]	      91 ┤▊                               │
	[  64,  127]	      57 ┤▍                               │
	[ 128,  255]	     207 ┤█▋                              │
	...
	[ 512,   1k)	      10 ┤▏                               │
	[  1k,   2k)	      21 ┤▏                               │

END
```

riscv64:

```
$ uname -a
Linux unmatch5 6.0.7-gentoo-riscv64 #1 SMP Sun Nov 6 11:07:10 CST 2022 riscv64 GNU/Linux
$ sudo ply -c 'sleep 5' kretprobe.ply
BEGIN

@:
{ size    }: 
	[   0,    1]	       1 ┤██▎                             │
	...
	[   8,   15]	       2 ┤████▋                           │
	...
	[ 128,  255]	       1 ┤██▎                             │
	[ 256,  511]	       2 ┤████▋                           │
	[ 512,   1k)	       8 ┤██████████████████▎             │

END
```

qemu riscv64:

```
$ uname -a
Linux ubuntu 5.15.53 #3 SMP Wed Oct 19 08:35:10 UTC 2022 riscv64 riscv64 riscv64 GNU/Linux
$ sudo ply -c 'sleep 5' kretprobe.ply
info: creating kallsyms cache
BEGIN

@:
{ size    }: 
	[   0,    1]	       2 ┤██████████▋                     │
	...
	[  32,   63]	       1 ┤█████▍                          │
	...
	[ 512,   1k)	       2 ┤██████████▋                     │
	...
	[  2k,   4k)	       1 ┤█████▍                          │

END
```

File: include/ply/syscall.h [hunk 2]
```diff
@@ -34,4 +34,15 @@ int bpf_map_next  (int fd, void *key, void *next_key);
...
 		    int cpu, int group_fd, unsigned long flags);
 
+#ifndef offsetof
+#define offsetof(TYPE, MEMBER) ((size_t) &((TYPE *)0)->MEMBER)
+#endif
+
+#ifndef offsetofend
+#define offsetofend(TYPE, FIELD) \
+	(offsetof(TYPE, FIELD) + sizeof(((TYPE *)0)->FIELD))
+#endif
```
Comment (wkz) [inline include/ply/syscall.h:44]: We shouldn't need this. Please remove.

See comment on `bpf_prog_test_run`

File: src/libply/aux/syscall.c [hunk 3]
```diff
@@ -59,6 +57,16 @@ int bpf_map_create(enum bpf_map_type type, int key_sz, int val_sz, int entries)
...
 
+int bpf_prog_test_run(int prog_fd)
+{
+	size_t attr_sz = offsetofend(union bpf_attr, test);
+	union bpf_attr attr;
+
+	memset(&attr, 0, attr_sz);
+	attr.test.prog_fd = prog_fd;
+
+	return syscall(__NR_bpf, BPF_PROG_TEST_RUN, &attr, attr_sz);
```
Comment (wkz) [inline src/libply/aux/syscall.c:68]: The last argument here should just be `sizeof(attr)`. Its only purpose is to let the kernel work with older binaries that were linked to old headers where new members didn't exist yet.

File: src/libply/libply.c [hunk 4]
```diff
@@ -360,6 +377,63 @@ static int ply_load_bpf(struct ply *ply)
...
+
+	close(fd);
+	has_result = 1;
+
+	if (!can_prog_test_run_)
+		_w("The current environment does not support BEGIN/END\n");
+
+	return can_prog_test_run_;
+}
+
```
Comment (wkz) [inline src/libply/libply.c:436]: I don't think we need this. We already know that `BPF_PROG_TEST_RUN` is defined, so this should all work. If something isn't working, `ply -T` should already be able to pinpoint the issue.

**Review by wkz — COMMENTED** [review]
Awesome work, this is getting reeeeally close now :+1:

File: include/ply/syscall.h (hunk 2 above)
Comment (xmzzz) [reply include/ply/syscall.h]: OK, Fixed in [93bbc1b](https://github.com/iovisor/ply/pull/82/commits/93bbc1b022b4c2587e0b32aa3fd52bb752da5e2a) .

File: src/libply/aux/syscall.c (hunk 3 above)
Comment (xmzzz) [reply src/libply/aux/syscall.c]: I see, that would be cool. Fixed in [93bbc1b](https://github.com/iovisor/ply/pull/82/commits/93bbc1b022b4c2587e0b32aa3fd52bb752da5e2a) .

File: src/libply/libply.c (hunk 4 above)
Comment (xmzzz) [reply src/libply/libply.c]: Indeed, fixed in [93bbc1b](https://github.com/iovisor/ply/pull/82/commits/93bbc1b022b4c2587e0b32aa3fd52bb752da5e2a) .
Please take a look, thanks.
Tested after modification, the result is normal:
```
$ sudo ply -T
Verifying kernel config (/boot/config-5.15.0-53-generic)... OK
Ensuring that debugfs is mounted... OK
Verifying kprobe... OK
Verifying tracepoint... OK
Verifying special... OK
Verifying interval... OK
```

**Comment by jsmolic** [comment]
@wkz Thanks for merging, any chance we could get a new release? It would make things easier for us as downstream Gentoo package maintainers to add riscv support for ply in Gentoo

**Comment by wkz** [comment]
@jsmolic yeah it was on my TODO, thanks for the nudge :smile: 2.3.0 is now available.

**Pull request merged**